### ISSUE TITLE
Feature: clang CUDA (device compilation) support

### ISSUE THREAD

**Issue by MatthiasKohl** [NVIDIA/libcudacxx#238]
This issue is similar to this one in cuCollections: https://github.com/NVIDIA/cuCollections/issues/128

While working on RAFT (primitives for RADPIS, PR: https://github.com/rapidsai/raft/pull/424), I noticed that libcudacxx doesn't support having its device code compiled with clang.

It would be great to support this, if possible.
If test coverage is good enough, it's usually good enough to simply transform `nvcc` commands to clang for which I have a somewhat usable script in the above mentioned PR for RAFT, and apply those commands to the entire test suite.

**Comment by wmaxey** [comment]
We'll place this on a milestone once we get gpuCI, that's a prerequisite to this as we don't have the capacity on DVS.

**Comment by gonzalobg** [comment]
This was supported a year ago. @MatthiasKohl what error are you encountering?

**Comment by MatthiasKohl** [comment]
Responded with PM, the error comes from using [cuCollections](https://github.com/NVIDIA/cuCollections) and in particular, the static map implementation.

**Comment by upsj** [comment]
I am having the same kind of issue trying to compile [cuDF](https://github.com/rapidsai/cudf) with clang as device compiler. The specific error I'm observing occurs in type_traits:
<details>

```
In file included from /home/tribizel/cudf/cpp/build/_deps/libcudacxx-src/include/cuda/std/limits:16:
In file included from /home/tribizel/cudf/cpp/build/_deps/libcudacxx-src/include/cuda/std/type_traits:24:
/home/tribizel/cudf/cpp/build/_deps/libcudacxx-src/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions
false_type __sfinae_test_impl(...);
           ^
/home/tribizel/cudf/cpp/build/_deps/libcudacxx-src/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions
    template <class _Tp> _LIBCUDACXX_INLINE_VISIBILITY static __two __test(...);
                                                                    ^
/home/tribizel/cudf/cpp/build/_deps/libcudacxx-src/include/cuda/std/detail/libcxx/include/type_traits:1171:5: error: CUDA device code does not support variadic functions
    __any(...);
    ^
/home/tribizel/cudf/cpp/build/_deps/libcudacxx-src/include/cuda/std/detail/libcxx/include/type_traits:1848:16: error: CUDA device code does not support variadic functions
   static void __test(...);
               ^
/home/tribizel/cudf/cpp/build/_deps/libcudacxx-src/include/cuda/std/detail/libcxx/include/type_traits:2272:18: error: CUDA device code does not support variadic functions
    static __two __test (...);
                 ^
/home/tribizel/cudf/cpp/build/_deps/libcudacxx-src/include/cuda/std/detail/libcxx/include/type_traits:4332:16: error: CUDA device code does not support variadic functions
  static __nat __try_call(...);
```
</details>

**Comment by maddyscientist** [comment]
For clang you probably need to enable the use of variadic functions with `-fcuda-allow-variadic-functions`

**Comment by Artem-B** [comment]
A word of caution. Clang currently does not have support for GPU-side variadic functions but in this case it's benign, as this is only a declaration used as a low-priority overload candidate for which we will not need to generate any code. That said, when the option is in effect, it may lead to a miscompilation or compiler crash if you try to compile an actual variadic function for a GPU. There's been some progress on this (https://github.com/llvm/llvm-project/issues/58410). LLVM support has been added recently, but it's not been plumbed through to the CUDA front-end yet) 

If you run into clang issues, please do file an issue in LLVM's repo. libcudacxx is getting more users and it would be useful to know what features are actually needed. Otherwise we end up in a bit of a catch-22 situation -- people can't use clang for CUDA compilation because feature X is missing, and the feature X is missing because nobody asks for it.

**Comment by Artem-B** [comment]
> will just emit garbage

I do not think so. You will probably run into and error if there's a real attempt to use varargs via va_start/va_arg. E.g. https://godbolt.org/z/3qa5dn88c

Most of the required functionality is already in place, we're missing a bit of glue to tie it all together.
